src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php

```diff
@@ -0,0 +1,175 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Loader\PhpFileLoader;
+use Symfony\Component\ExpressionLanguage\Expression;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @method IteratorArgument iterator(array $values)
+ */
+class ContainerConfigurator extends AbstractConfigurator
+{
+    const FACTORY = 'container';
+
+    private $file;
+    private $container;
+    private $loader;
+
+    public function __construct($file, ContainerBuilder $container, PhpFileLoader $loader, &$instanceof)
+    {
+        $this->file = $file;
+        $this->container = $container;
+        $this->loader = $loader;
+        $this->instanceof = &$instanceof;
+    }
+
+    final public function extension($namespace, array $config)
+    {
+        if (!$this->container->hasExtension($namespace)) {
+            $extensions = array_filter(array_map(function ($ext) { return $ext->getAlias(); }, $this->container->getExtensions()));
+            throw new InvalidArgumentException(sprintf(
+                'There is no extension able to load the configuration for "%s" (in %s). Looked for namespace "%s", found %s',
+                $namespace,
+                $this->file,
+                $namespace,
+                $extensions ? sprintf('"%s"', implode('", "', $extensions)) : 'none'
+            ));
+        }
+
+        $this->container->loadFromExtension($namespace, PhpFileLoader::processValue($config));
+    }
+
+    final public function import($resource, $type = null, $ignoreErrors = false)
+    {
+        $this->loader->import($resource, $type, $ignoreErrors, $this->file);
+        $this->loader->setCurrentDir(dirname($this->file));
+    }
+
+    /**
+     * @return ParametersConfigurator
+     */
+    final public function parameters()
+    {
+        return new ParametersConfigurator($this->container);
+    }
+
+    /**
+     * @return ServicesConfigurator
+     */
+    final public function services()
+    {
+        return new ServicesConfigurator($this->container, $this->loader, $this->instanceof);
+    }
+
+    /**
+     * Creates a service reference.
+     *
+     * @param string $id
+     *
+     * @return ReferenceConfigurator
+     */
+    final public function reference($id)
+    {
+        return ref($id);
+    }
+
+    /**
+     * Creates an inline service.
+     *
+     * @param string|null $class
+     *
+     * @return InlineServiceConfigurator
+     */
+    final public function inline($class = null)
+    {
+        return inline($class);
+    }
+
+    /**
+     * Creates a lazy iterator.
+     *
+     * @param ReferenceConfigurator[] $values
+     *
+     * @return IteratorArgument
+     */
+    final protected function setIterator(array $values)
+    {
+        return iter($values);
+    }
```

Comment:
I'm sure there's a reason, but I can't see why this is protected and relies on the magic method to become exposed publicly. Is that needed?

Reply:
iterator is a reserved keyword.

Reply:
Even as a method name? If so, is this the only reason the magic method in the base class exists? Or does that also serve other purposes? I apologize - I'm reviewing in chunks...

Reply:
Starting with php7, reserved keywords are allowed as method names, so we can remove the magic method on master.
About "iterator", my mistake: it's not a reserved keyword.